### ISSUE TITLE
[Code Health] The DFX drain path acks a ready entry before it can deliver it, so a failed resolve loses records uncharged and strands the buffer

### ISSUE THREAD

**Issue by ChaoWao** [hw-native-sys/simpler#2335]
Related: #995 (tracking), #997 (closed — device-side capacity drops), #2147 (closed — per-lane starvation recovery), #1161 (closed — throughput drops). All three of those concern drops the **device** takes when it cannot get a buffer, which are charged to `dropped_record_count`. This issue is the opposite direction: a drop the **host** takes on an entry it has already acknowledged, which is charged to nothing.

Observed in the wild on the DFX smokes (see the log excerpt in #2181's second comment), but filed as Code Health rather than a bug because it is not independently reproducible on demand — it needs a transient resolve/copy failure, and the causes of those have their own issues (#2220 / #2237 / #2206 / #2320). This is the framework behaviour that turns any of them from a recoverable hiccup into unaccounted loss.

### Category

Robustness (potential edge-case failure)

### Component

Host Runtime

### Description

`ProfilerAlgorithms` splits draining a device ready entry into two steps, and the commit point sits at the end of the first one rather than after delivery.

`try_pop_aicpu_entry()` advances `header->queue_heads[q]` and **writes the new head back to the device** before it returns `true`. That write-back is the acknowledgment: the device is now free to reuse the ready slot. The function treats it as the commit point and is careful about it — if the write-back fails it restores `old_head` and returns `false`.

`process_entry()` then runs with no such care. Three separate failure paths return early, and each one is a bare `return`:

```cpp
auto site_opt = Module::resolve_entry(mgr.shared_mem_host(), header, q, entry);
if (!site_opt.has_value()) return;                       // profiler_base.h:448

site.info.host_buffer_ptr = mgr.resolve_host_ptr(site.info.dev_buffer_ptr);
if (site.info.host_buffer_ptr == nullptr) {
    // resolve_host_ptr already logged. Drop rather than deliver null.
    return;                                              // profiler_base.h:454
}
if (mgr.copy_buffer_from_device(...) != 0) {
    LOG_ERROR(...);
    return;                                              // profiler_base.h:462
}
```

The comment at `profiler_base.h:474` states the ordering plainly — *\"The device ready entry was already acknowledged by `try_pop_aicpu_entry()`\"* — so each of those returns has two consequences, and neither is recoverable.

**1. The records are lost and not accounted for.** Nothing re-queues the entry and nothing increments any host-side drop counter. `reconcile_counters()` computes `silent_loss = device_total - (collected + dropped_device)`, where `dropped_device` is read from device state only, so a host-side drop is structurally invisible to it and can only ever appear as silent loss:

```text
BufferPoolManager: no host mapping for dev_ptr=0x120000000000
PERF count mismatch (collected=0 + dropped=0 != device_total=5, silent_loss=5)
```

Those two lines are the same event one layer apart — `resolve_host_ptr` logging its failure at `buffer_pool_manager.h:865`, then reconcile reporting the gap it caused. The reconcile warning is the only evidence the entry ever existed, and it names a number rather than a cause.

**2. The buffer is stranded, so the loss compounds.** A buffer only re-enters circulation via `consume()`, which calls `on_buffer_collected()` and then `manager_.notify_copy_done()` (`profiler_base.h:1309-1318`). `consume()` is reached only through `wait_push_to_ready()` at `profiler_base.h:478` — after all three early returns. Note also that the free_queue top-up at `profiler_base.h:469` sits after them too, so the failing path neither delivers the popped buffer nor refills the lane it came from.

So each failed entry permanently removes one buffer from that kind's pool. The pool only shrinks. Once it is dry, the lane starves and drops more — which is the escalation path that makes `collected=0` out of `device_total=5` a plausible outcome of a single initial hiccup rather than five independent ones. I have not confirmed that this is what produced any specific `collected=0` report; I am noting that the structure permits it, which matters when triaging the family above.

Why this is worth fixing separately from the individual causes: the smoke tests assert on record completeness with no tolerance, so any loss here is a red CI job, and the only diagnostic it leaves is a count. Every fix in the #2220 / #2237 / #2206 / #2320 family removes one *cause* of a transient failure; none of them changes what happens when the next one occurs.

### Location

- `src/common/platform/include/host/profiler_base.h:381-430` — `try_pop_aicpu_entry()`; the head advance and device write-back (the ack) are at `:419-428`
- `src/common/platform/include/host/profiler_base.h:445-479` — `process_entry()`; the three uncharged early returns are at `:448`, `:454`, `:462`
- `src/common/platform/include/host/profiler_base.h:469` — free_queue top-up, after all three returns
- `src/common/platform/include/host/profiler_base.h:478` — `wait_push_to_ready()`, the only route to `consume()`
- `src/common/platform/include/host/profiler_base.h:1309-1318` — `consume()`; `notify_copy_done()` is what recycles the buffer
- `src/common/platform/include/host/buffer_pool_manager.h:865` — `resolve_host_ptr()` failure log
- `src/common/platform/shared/host/chip_swimlane_collector.cpp:745` and `:806` — the `silent_loss` derivation and its warning
- Same shape reaches every collector through the shared base: `chip_swimlane_collector.cpp`, `dep_gen_collector.cpp`, `args_dump_collector.cpp`, `scope_stats_collector.cpp`, and both `pmu_collector.cpp` copies

Reference commit: `b8e9dfe9871ee88c0ab12e0cb90d303361fa1951`

### Proposed Fix

Three options, cheapest first. They are not exclusive — (1) is worth doing regardless.

1. **Charge the loss.** Add a host-side dropped counter that `reconcile_one()` folds into `accounted`, so these become classified drops instead of silent loss. Purely additive, no protocol change, and it converts an unattributable number into a named failure with a call site. This alone would have told #2181 and #2246 which of the two mechanisms they were looking at.

2. **Do not strand the buffer.** On the `resolve_host_ptr` / `copy_buffer_from_device` paths the device pointer is in hand, so the buffer can be returned to its recycled lane before returning. That stops the pool from shrinking and removes the compounding, without touching the ack protocol. The `resolve_entry` path is harder — it is precisely the case where the site could not be determined — so it may have to stay a leak and just be counted under (1).

3. **Move the ack after delivery.** The structurally correct fix, and the expensive one: it changes when the device may reuse a ready slot, so it needs a re-examination of the SPSC invariant between the drain shard and the AICPU producer, and the rollback would have to be safe against a producer that has already observed the advanced head. **Not something to bundle into a fix for any of the issues above** — worth its own design discussion.

Suggest landing (1) first so the next occurrence in the #2220 family arrives with a cause attached rather than a count.

### Priority

Medium (minor risk, should fix in next few releases)

**Comment by ChaoWao** [comment]
## Not the same defect as #2346 — flagging before the resemblance closes this

#2346 (*Fix: name a DFX publication the device rejected instead of discarding it*) shares this issue's file set almost exactly — `profiler_base.h` plus every collector — and both are about a DFX loss going unreported. They are different defects, and closing this as a duplicate would drop the one described here.

| | #2346 | this issue |
| --- | --- | --- |
| direction | host → device | device → host |
| path | the twelve `write_range_to_device` config/counter pushes | the drain path, `ProfilerAlgorithms::process_entry` |
| what is lost | a configuration write that the manager's bounds check rejected | a ready entry the host **already acknowledged** |
| how it hides | return value cast to `(void)` | no counter is incremented at all, so `reconcile_counters` can only report it as `silent_loss` |
| buffer | unaffected | permanently out of circulation |

Checked rather than inferred: #2346's only hunk in `profiler_base.h` is `@@ -980,6 +980,36 @@` (adding `publish_field`). `process_entry` is at `:445-479`, and its three early returns at `:448`, `:454` and `:462` are untouched by it. #2326's hunks in that header are at 124 / 249 / 744 / 763 / 779 / 819 / 981 and also miss that range.

So after both land, a *rejected config write* becomes attributable and the base-pointer disagreement that causes it is fixed — but an **acknowledged ready entry dropped during resolve/copy** still increments nothing and still strands its buffer. That is what this issue is for.

## Deliberately not fixing it yet

`profiler_base.h` currently has two open PRs against it (#2326, which deletes 49 lines from `start()`, and #2346), and both add cases to `tests/ut/cpp/common/test_profiler_base.cpp`. A third concurrent change to the same header — and the natural home for a new counter member is right at the line both of them touch — would create a three-way merge for a diagnostic improvement that is not urgent.

Better after those land. The shape has not changed from the issue body: option (1) alone (charge the loss so `reconcile_one` folds it into `accounted`) is additive and independent of the ack ordering, and is what makes the *next* occurrence name its own cause.
